src/Event.cc

```diff
@@ -204,3 +211,29 @@ void EventMgr::Describe(ODesc* d) const
 		d->NL();
 		}
 	}
+
+void EventMgr::Process()
+	{
+	// If we don't have a source, or the source is closed, or we're
+	// reading live (which includes pseudo-realtime), advance the time
+	// here to the current time since otherwise it won't move forward.
+	iosource::PktSrc* pkt_src = iosource_mgr->GetPktSrc();
+	if ( ! pkt_src || ! pkt_src->IsOpen() || reading_live )
+		net_update_time(current_time());
+
+	queue_flare.Extinguish();
```

Comment:
It comes down to one of these two cases:

With `Extinguish()` in `Drain()`:

- Loop is interrupted by something other than `EventMgr` (meaning the flare isn't lit)
- New events are added as packets are processed
- `Fire()` is called because head was null
- `Drain()` is called
- Head is set back to null
- No new events are added by events processed in `Drain()`
- `Extinguish()` is called because we had events originally and head is now null

In this case, the Fire/Extinguish pair is unnecessary.

`Extinguish()` in `Process()`:

- Loop is interrupted by something other than `EventMgr` (no events in queue)
- New events are added as packets are processed
- `Fire()` is called because head was null
- `Drain()` is called
- Head is set back to null
- No new events are added by events processed in `Drain()`

In this case the `kevent()` gets interrupted here for no reason by the flare, since there's no events to be processed by `Drain()`.

Another concern is that moving `Extinguish()` to `Drain()` has a case where it's possible to continually call `Fire()` as part of `Drain()` without ever getting to extinguish it because we're always adding new events while processing old ones. I could see that being a memory leak eventually as the pipe fills up.